### DISCUSSION TITLE
Is RegionAdapter a singleton for different instances of the same control?

### DISCUSSION THREAD

**Discussion by codingdna2** [PrismLibrary/Prism#2460] [WPF] [answered]
Hi, not sure is the right place to ask a question, so excuse me in advance if it's not.

**Background**

I implemented a custom adapter for `DXTabControl` from DevExpress which supports tabs tear-off in a new window. I followed all your courses on pluralsight and went on and implemented `RegionAdapterBase<DXTabControl>` in order to add elements to the control itself and a `RegionManagerAwareBehavior` to set the right `IRegionManager` inside my ViewModels after creating a Scoped region.

**Problem**

When I've implemented the `RegionAdapterBase`, I've followed some examples and placed some class properties until I've realized this variables were overwritten when a second call to Adapt was made (Region and Target).

**Comment by brianlagunas** [comment, marked as answer]
RegionAdapters are not scoped to a specific region instance. They apply to all regions of that specific control type.

**Reply by codingdna2** [reply]
Hi Brian, thanks for the fast response. I see a behavior is created for each region, shall I try to move my logic in there? Any suggestion is highly appreciated :)

**Reply by brianlagunas** [reply]
You will need a combination of an adapter and a behavior. If you get stuck, I do offer consulting services. You can DM me on twitter if you want more information about those services.

**Reply by codingdna2** [reply]
Hi Brian, that's an option I'm considering... but I've taken the path of pluralsight and I've to try some more.. (surely if I knew before I would have taken a different decision). Hopefully you can reply me one more time before I drop everything and start crying ;)

EDIT:
I now get an error when I tear-off a view containing a sub-region. The `RegionManagerRegistrationBehavior` detects that `targetElement` regionManager has changed (ref. `TryRegisterRegion`, line 110), remove the region from the `attachedRegionManager` but here things became recursive... Inside `RegionManager.Remove(string regionName)` it calls `UpdateRegions` which again call `TryRegisterRegion` which add the Region to the new RegionManager. Now when stack trace unfolds, the caller is also trying to add the same Region.

**Reply by codingdna2** [reply]
I've updated the previous comment... Is this a bug? 
I've tested changing the second `if` with an `else if` and it solve my issue but I'm not sure about other implications.

```cs
private void TryRegisterRegion()
{
    DependencyObject targetElement = this.HostControl;
    if (targetElement.CheckAccess())
    {
        IRegionManager regionManager = this.FindRegionManager(targetElement);

        IRegionManager attachedRegionManager = this.GetAttachedRegionManager();

        if (regionManager != attachedRegionManager)
        {
            if (attachedRegionManager != null)
            {
                this.attachedRegionManagerWeakReference = null;
                attachedRegionManager.Regions.Remove(this.Region.Name);    // <= RECURSION HAPPENING HERE
            }
            else if (regionManager != null)       // <= CHANGED THIS
            {
                this.attachedRegionManagerWeakReference = new WeakReference(regionManager);
                regionManager.Regions.Add(this.Region);
            }
        }
    }
}
```

Edit: of course this can't be the solution but I guess is enough to explain the issue. Next week I'll try to create a minimal repro or a unit test.

**Reply by codingdna2** [reply]
Opened issue #2471
